gpumon/gpumon_interface.ml

```diff
@@ -12,60 +12,147 @@
  *)
 
+open Rpc
+open Idl
+
 let service_name = "gpumon"
 let queue_name = Xcp_service.common_prefix ^ service_name
 let xml_path = "/var/xapi/" ^ service_name
 
+(** Uninterpreted string associated with the operation *)
 type debug_info = string
+[@@deriving rpcty]
+
+(* Domain ID of VM *)
 type domid = int
+[@@deriving rpcty]
+
+(** Reason for incompatibility *)
+type incompatibility_reason =
+  | Host_driver
+  | Guest_driver
+  | GPU
+  | Other
+[@@deriving rpcty]
 
-type incompatibility_reason = Host_driver | Guest_driver | GPU | Other
-type compatibility = Compatible | Incompatible of incompatibility_reason list
+(** Boolean: compatible? *)
+type compatibility =
+  | Compatible
+  | Incompatible of incompatibility_reason list
+[@@deriving rpcty]
 
+(** PCI identifier of physical GPU *)
 type pgpu_address = string
+[@@deriving rpcty]
+
+(** Metadata of Nvidia physical GPU *)
 type nvidia_pgpu_metadata = string
+[@@deriving rpcty]
+
+(** Metadata of Nvidia virtual GPU *)
 type nvidia_vgpu_metadata = string
+[@@deriving rpcty]
+
+(** List of Nvidia virtual GPU metadata records *)
+type nvidia_vgpu_metadata_list = nvidia_vgpu_metadata list
+[@@deriving rpcty]
+
+
+(** Error wrapper *)
+type gpu_errors =
+  | NvmlInterfaceNotAvailable
+  (** Exception raised when gpumon is unable to load the nvml nvidia library *)
+  | NvmlFailure of string
+  (** Exception raised by the c bindings to the nvml nvidia library*)
+  | Gpumon_failure
+  (** Default exception raised upon daemon failure *)
+[@@default Gpumon_failure]
+[@@deriving rpcty]
+
+exception Gpumon_error of gpu_errors
+
+(** Error handler *)
+module GpuErrors = Error.Make(struct
+    type t = gpu_errors
+    let t = gpu_errors
+  end)
+let gpu_err = GpuErrors.error
+
+(** Functor to autogenerate API calls *)
+module RPC_API(R : RPC) = struct
+  open R
+
+  let param = Param.mk
 
-(** Exception raised when gpumon is unable to load the nvml nvidia library *)
-exception NvmlInterfaceNotAvailable
-(** Exception raised by the c bindings to the nvml nvidia library*)
-exception NvmlFailure of string
+  let description =
+    Interface.{ name = "Gpumon"
+              ; namespace = None
+              ; description =
+                  [ "This interface is used by Xapi and Gpumon to monitor "
+                  ; "physical and virtual GPUs."]
+              ;
+                version=(1,0,0)
+              }
 
+  let implementation = implement description
+
+  (** common API call parameters *)
+
+  let debug_info_p = param ~description:
+      ["Uninterpreted string used for debugging."]
+      debug_info
+
+  let domid_p = param ~description:
+      ["Domain ID of the VM in which the vGPU(s) is running."]
+      domid
+
+  let pgpu_address_p = param ~description:
+      ["PCI bus ID of the pGPU in which the VM is currently running"
+      ;"in the form `domain:bus:device.function` PCI identifier."]
+      pgpu_address
+
+  let nvidia_pgpu_metadata_p = param ~description:
```

Comment:
It could make sense, now that it is possible, to move all these `nvidia_` types, errors and functions to an `Nvidia` submodule dropping the prefix and encapsulating things more appropriately.